internal/logging/context.go

```diff
@@ -15,39 +17,107 @@ import (
 
 type loggerContextKey struct{}
 
-var globalLogger *log.Entry
+var (
+	globalLogger logr.Logger
+
+	// programLevel allows for setting the logging level dynamically
+	// via SetLevel().
+	programLevel = new(slog.LevelVar)
+)
+
+const (
+	LevelTrace = slog.Level(-2)
+	LevelDebug = slog.Level(-1)
+	LevelInfo  = slog.Level(0)
+	LevelError = slog.Level(8)
+
+	// These constants contain logging level strings,
+	// purely for the performance benefit.
+	TRACE = "TRACE"
+	DEBUG = "DEBUG"
+	INFO  = "INFO"
+	ERROR = "ERROR"
+)
 
 func init() {
-	globalLogger = log.New().WithFields(nil)
-	level, err := log.ParseLevel(os.GetEnv("LOG_LEVEL", "INFO"))
+	level, err := parseLevel(os.GetEnv("LOG_LEVEL", "INFO"))
 	if err != nil {
 		panic(err)
 	}
-	globalLogger.Logger.SetLevel(level)
+	programLevel.Set(level)
+
+	opts := &slog.HandlerOptions{
+		Level: programLevel,
+		ReplaceAttr: func(groups []string, a slog.Attr) slog.Attr {
+			// Customize the name of the level key and the output string,
+			// including custom level values. Meaning:
+			// log.V(1).Info() becomes level=DEBUG instead of level=DEBUG+3
+			// log.V(2).Info() becomes level=TRACE instead of level=DEBUG+2
+			if a.Key == slog.LevelKey {
+				// Handle custom level values.
+				level := a.Value.Any().(slog.Level)
+
+				switch {
+				case level < LevelDebug:
+					a.Value = slog.StringValue(TRACE)
+				case level < LevelInfo:
+					a.Value = slog.StringValue(DEBUG)
+				case level < LevelError:
+					a.Value = slog.StringValue(INFO)
+				default:
+					a.Value = slog.StringValue(ERROR)
+				}
+			}
+			return a
+		},
+	}
+
+	globalLogger = logr.FromSlogHandler(slog.NewTextHandler(os.Stderr, opts))
+
 	SetKLogLevel(os.GetEnvInt("KLOG_LEVEL", 0))
 
-	runtimelog.SetLogger(logrusr.New(globalLogger))
+	runtimelog.SetLogger(globalLogger)
 }
 
 // ContextWithLogger returns a context.Context that has been augmented with
-// the provided log.Entry.
-func ContextWithLogger(ctx context.Context, logger *log.Entry) context.Context {
+// the provided logr.Logger.
+func ContextWithLogger(ctx context.Context, logger logr.Logger) context.Context {
 	return context.WithValue(ctx, loggerContextKey{}, logger)
 }
 
-// LoggerFromContext extracts a *log.Entry from the provided context.Context and
-// returns it. If no *log.Entry is found, a global, error-level *log.Entry is
-// returned.
-func LoggerFromContext(ctx context.Context) *log.Entry {
+// LoggerFromContext extracts a logr.Logger from the provided context.Context and
+// returns it. If no logr.Logger is found, a global logr.Logger is returned.
+func LoggerFromContext(ctx context.Context) logr.Logger {
```

Comment:
Wondering if at this point, it wouldn't simply be better to make use of the tooling that comes with controller-runtime: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/log#pkg-index

Reply:
@hiddeco I've had the same thought previously and have held off so-far because our `logging` package also does some logger initialization based on environment variables. If we stop relying on that, we need to come up with another way of doing that. I'm open to it, but think it should be handled in a follow-up.